mvebu: add initial support for Globalscale ESPRESSObin-Ultra

v-vid opened the pull request:
This patch adds support for Globalscale ESPRESSObin-Ultra. Device uses
the same Armada-3720 SoC with extended hardware support.

- SoC: Armada-3720
- RAM: 1 GB DDR4
- Flash: 4MB SPI NOR (mx25u3235f) + 8 GB eMMC
- Ethernet: Topaz 6341 88e6341 (4x GB LAN + 1x WAN with 30W PoE)
- WiFI: 2x2 802.11ac Wi-Fi marvell (88w8997 PCIe+USB)
- 1x USB 2.0 port
- 1x USB 3.0 port
- 1x microSD slot
- 1x mini-PCIe slot (USB [with nano-sim slot])
- 1x mini-USB debug UART
- 1x RTC Clock and battery
- 1x reset button
- 1x power button
- 4x LED (RGBY)
- Optional 1x M.2 2280 slot

** Installation **

1. Copy dtb from build_dir to bin/ and run tftpserver there:
  $ cp ./build_dir/target-aarch64_cortex-a53_musl/linux-mvebu_cortexa53/\
linux-5.4.65/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dtb \
bin/targets/mvebu/cortexa53/
  $ in.tftpd -L -s bin/targets/mvebu/cortexa53/

2. Connect to the device UART via microUSB port on the back side and power on the device.
3. Power on the device and hit any key to stop the autoboot.
4. Set serverip (host IP) and ipaddr (any free IP address on the same subnet), e.g:
  $ setenv serverip 192.168.1.10 # Host
  $ setenv ipaddr 192.168.1.15 # Device
5. Ping server to confirm network is working:
  $ ping $serverip
  Using neta@30000 device
  host 192.168.1.15 is alive
6. Tftpboot the firmware:
  $ tftpboot $kernel_addr_r openwrt-mvebu-cortexa53-globalscale_espressobin-ultra-initramfs-kernel.bin
  $ tftpboot $fdt_addr_r armada-3720-espressobin-ultra.dtb
7. Set the console and boot the image:
  $ setenv bootargs $console
  $ booti $kernel_addr_r - $fdt_addr_r
8. Once the initramfs is booted, transfer openwrt-mvebu-cortexa53-globalscale_espressobin-ultra-squashfs-sdcard.img.gz
   to /tmp dir on the device.
9. Gunzip and dd the image:
  $ gunzip /tmp/openwrt-mvebu-cortexa53-globalscale_espressobin-ultra-squashfs-sdcard.img.gz
  $ dd if=/tmp/openwrt-mvebu-cortexa53-globalscale_espressobin-ultra-squashfs-sdcard.img of=/dev/mmcblk0 && sync
10. Reboot the device.

Commit:
mvebu: add initial support for Globalscale ESPRESSObin-Ultra

This patch adds support for Globalscale ESPRESSObin-Ultra. Device uses
the same Armada-3720 SoC with extended hardware support.

- SoC: Armada-3720
- RAM: 1 GB DDR4
- Flash: 4MB SPI NOR (mx25u3235f) + 8 GB eMMC
- Ethernet: Topaz 6341 88e6341 (4x GB LAN + 1x WAN with 30W PoE)
- WiFI: 2x2 802.11ac Wi-Fi marvell (88w8997 PCIe+USB)
- 1x USB 2.0 port
- 1x USB 3.0 port
- 1x microSD slot
- 1x mini-PCIe slot (USB [with nano-sim slot])
- 1x mini-USB debug UART
- 1x RTC Clock and battery
- 1x reset button
- 1x power button
- 4x LED (RGBY)
- Optional 1x M.2 2280 slot

** Installation **

Copy dtb from build_dir to bin/ and run tftpserver there:
$ cp ./build_dir/target-aarch64_cortex-a53_musl/linux-mvebu_cortexa53/
linux-5.4.65/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dtb
bin/targets/mvebu/cortexa53/
$ in.tftpd -L -s bin/targets/mvebu/cortexa53/

Connect to the device UART via microUSB port on the back side and power on the device.

Power on the device and hit any key to stop the autoboot.

Set serverip (host IP) and ipaddr (any free IP address on the same subnet), e.g:
$ setenv serverip 192.168.1.10 # Host
$ setenv ipaddr 192.168.1.15 # Device

Ping server to confirm network is working:
$ ping $serverip
Using neta@30000 device
host 192.168.1.15 is alive

Tftpboot the firmware:
$ tftpboot $kernel_addr_r openwrt-mvebu-cortexa53-globalscale_espressobin-ultra-initramfs-kernel.bin
$ tftpboot $fdt_addr_r armada-3720-espressobin-ultra.dtb

Set the console and boot the image:
$ setenv bootargs $console
$ booti $kernel_addr_r - $fdt_addr_r

Once the initramfs is booted, transfer openwrt-mvebu-cortexa53-globalscale_espressobin-ultra-squashfs-sdcard.img.gz
to /tmp dir on the device.

Gunzip and dd the image:
$ gunzip /tmp/openwrt-mvebu-cortexa53-globalscale_espressobin-ultra-squashfs-sdcard.img.gz
$ dd if=/tmp/openwrt-mvebu-cortexa53-globalscale_espressobin-ultra-squashfs-sdcard.img of=/dev/mmcblk0 && sync

Reboot the device.

Signed-off-by: Vladimir Vid <vladimir.vid@sartura.hr>

adschm commented:
Please add flashing instruction to your commit message.

What is "initial" in your proposal, and how can it be made "ready"?

adschm commented on target/linux/mvebu/cortexa53/base-files/lib/upgrade/platform.sh:
```diff
@@ -23,8 +24,10 @@ platform_check_image() {
 
 platform_do_upgrade() {
 	case "$(board_name)" in
+	globalscale,ccpe|\
```
Doesn't belong here?

v-vid replied:
sorry, CCPE was originalyl used as a code name and I managed to overlook this one while cleaning. Fixed now

adschm replied:
Still not fixed.

adschm commented on target/linux/mvebu/files-4.19/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dts:
```diff
@@ -0,0 +1,252 @@
+// SPDX-License-Identifier: (GPL-2.0+ OR MIT)
```
The file in files-4.19 can be removed, since the one in "files" will work for both kernels. (unless you have specific differences, then you would need files-4.19 and files-5.4 ...)

v-vid replied:
Removed dts from files-4.19.

adschm replied:
File is still there. Did you forget to push?

dhewg commented on package/boot/uboot-envtools/files/mvebu:
```diff
@@ -32,6 +32,9 @@ marvell,armada8040-mcbin-doubleshot|\
 	ubootenv_add_uci_config "/dev/mtd0" "0x3f0000" "0x10000" "0x10000" "1"
 	;;
+globalscale,espressobin-ultra)
+	ubootenv_add_uci_config "/dev/mtd2" "0x3f0000" "0x10000" "0x10000" "1"
```
Did you test this? The dts you're adding has:
```
			partition@0 {
				label = "u-boot";
				reg = <0x0 0x3e0000>;
			};
			partition@3e0000 {
				label = "hw-info";
				reg = <0x3e0000 0x10000>;
				read-only;
			};
			partition@3f0000 {
				label = "u-boot-env";
				reg = <0x3f0000 0x10000>;
			};
```

So that cannot work.

```suggestion
	idx="$(find_mtd_index u-boot-env)"
	[ -n "$idx" ] && \
		ubootenv_add_uci_config "/dev/mtd$idx" "0x0" "0x10000" "0x10000" "1"
```

v-vid replied:
I did, not sure why it would not work. This is the output from the device I have:

```
root@OpenWrt:~# cat /proc/mtd
dev:    size   erasesize  name
mtd0: 003e0000 00010000 "u-boot"
mtd1: 00010000 00010000 "hw-info"
mtd2: 00010000 00010000 "u-boot-env"

root@OpenWrt:~# cat /etc/fw_env.config
/dev/mtd2 0x3f0000 0x10000 0x10000 1

root@OpenWrt:~# fw_printenv
arch=arm
baudrate=115200
board=mvebu_armada-37xx
board_name=mvebu_armada-37xx
bootargs=console=ttyMV0,115200 earlycon=ar3700_uart,0xd0012000
bootcmd=mmc dev 0; load mmc 0:1 $loadaddr boot.scr; source $loadaddr
bootcmd_nfs=run get_images; run set_bootargs; booti $kernel_addr_r $ramdisk_addr_r $fdt_addr_r
bootdelay=2
console=console=ttyMV0,115200 earlycon=ar3700_uart,0xd0012000
cpu=armv8
...
```

dhewg replied:
Huh, 0x3f0000 is the offset, and none of those 3 partitions is big enough.
I got an error in that situation, which is why I added the suggested block in another PR, see #3405.
Can you please try my suggestion? That block is used by various boards already.

v-vid replied:
both cases work for me. I see that #3405 is still open, so I can remove this part for now and sent new PR just for this once your PR is merged. Would that be okay with you?

dhewg replied:
Sure, but that makes it a little incomplete, no? If you're fine with the suggested change, why not use it. This and the #3405 case would still be different blocks due to the fallback in my PR, so no need to wait for it.

v-vid replied:
@dhewg okay, opted with you suggestion and updated PR. I would appreciate if you could add espressobin-ultra under the globalscale boards case in #3405 if this one gets merged before, so that everything is unified there.

dhewg replied:
Thanks, and sounds like a plan, will do! :)

v-vid replied:
sorry, just now realized that I didn't push the change in the last PR. Fixed now.

dhewg replied:
;) Thanks!

dhewg replied:
#3405 has been merged, you can now just add the board name to the existing `case`

dhewg reviewed (commented):
Did you submit the dts to linux mainline?
There're some bits that dont' look like it would pass review as-is

v-vid commented:
> Please add flashing instruction to your commit message.
> 
@adschm Added instructions in the PR.

> What is "initial" in your proposal, and how can it be made "ready"?

By initial I meant base board support. Networking and most of peripherals is working at the moment, USB3 is currently broken as mentioned in the PR and I'm currently working on support for 2 88W8997 wifi cards with mwlwifi.

v-vid commented:
> Did you submit the dts to linux mainline?
> There're some bits that dont' look like it would pass review as-is

@dhewg Not yet, I'm still currently working on getting USB3 and wifi to work, after which I plan to cleanup dts if needed and push it to mainline linux.

tmn505 commented on target/linux/mvebu/patches-4.19/544-add-armada-3720-espressobin-ultra.dts.patch:
```diff
@@ -0,0 +1,9 @@
+--- a/arch/arm64/boot/dts/marvell/Makefile
++++ b/arch/arm64/boot/dts/marvell/Makefile
+@@ -1,5 +1,6 @@
+ # SPDX-License-Identifier: GPL-2.0
+ # Mvebu SoC Family
+ dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-db.dtb
+ dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-espressobin.dtb
++dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-espressobin-ultra.dtb
+ dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db.dtb
```
Remove this patch and following one, these are not necessary.

v-vid replied:
Removed both patches.

tmn505 commented on target/linux/mvebu/patches-5.4/600-add_espressobin_ultra.patch:
```diff
@@ -0,0 +1,10 @@
+--- a/arch/arm64/boot/dts/marvell/Makefile
++++ b/arch/arm64/boot/dts/marvell/Makefile
+@@ -2,6 +2,7 @@
+ # Mvebu SoC Family
+ dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-db.dtb
+ dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-espressobin.dtb
++dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-espressobin-ultra.dtb
+ dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-turris-mox.dtb
+ dtb-$(CONFIG_ARCH_MVEBU) += armada-3720-uDPU.dtb
+ dtb-$(CONFIG_ARCH_MVEBU) += armada-7040-db.dtb
```
Ditto.

dhewg commented:
> @dhewg Not yet, I'm still currently working on getting USB3 and wifi to work, after which I plan to cleanup dts if needed and push it to mainline linux.

Okay, but not all devices need to be working to submit for mainline inclusion. Maybe just remove usb3 and wifi nodes and submit, if you're lucky the reviewers find related bugs? :)

Any idea what the issue with usb3 and wifi is? Comphy related?

Oh, and are there any ultra specific patches required for u-boot, or is that still the same old u-boot based on the marvell fork?

v-vid commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dts:
```diff
@@ -0,0 +1,252 @@
+		led2 {
+			label = "led2";
+			gpios = <&gpionb 12 GPIO_ACTIVE_LOW>;
+			default-state = "off";
```
Updated.

v-vid commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dts:
```diff
@@ -0,0 +1,252 @@
+
+&avs {
+	status = "okay";
+};
```
Updated.

v-vid commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dts:
```diff
@@ -0,0 +1,252 @@
+
+&sdhci1 {
+	status = "disabled";
+};
```
Updated.

v-vid commented:
> > @dhewg Not yet, I'm still currently working on getting USB3 and wifi to work, after which I plan to cleanup dts if needed and push it to mainline linux.
> 
> Okay, but not all devices need to be working to submit for mainline inclusion. Maybe just remove usb3 and wifi nodes and submit, if you're lucky the reviewers find related bugs? :)

I know, however I did not have these issue on 4.19 so this bugs me

> 
> Any idea what the issue with usb3 and wifi is? Comphy related?
> 

Not really, as far as I understood there was some firmware related change between 4.19 and 5.4 but I didn't have time yet to look this up into more details. For usb it might be comphy related, for the wifi I'm pretty sure it's driver issue.

> Oh, and are there any ultra specific patches required for u-boot, or is that still the same old u-boot based on the marvell fork?

Same old Marvell fork. I also started working on u-boot/ATF support, I'll send new PR for this once I finish and test it.

dhewg commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dts:
```diff
@@ -0,0 +1,236 @@
+};
+
+&pcie0 {
+	status = "okay";
```
What about these?
```
	pinctrl-names = "default";
	pinctrl-0 = <&pcie_reset_pins &pcie_clkreq_pins>;
	reset-gpios = <&gpiosb XXX GPIO_ACTIVE_LOW>;
```
With XXX replaced with board specific number. Maybe that's related to your wifi fail?

dhewg commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dts:
```diff
@@ -0,0 +1,236 @@
+};
+
+&sata {
+	status = "okay";
```
What about these?
```
	phys = <&comphy2 0>;
	phy-names = "sata-phy";
```

dhewg commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dts:
```diff
@@ -0,0 +1,236 @@
+	mmc-ddr-1_8v;
+	mmc-hs400-1_8v;
+	marvell,pad-type = "fixed-1-8v";
+	status = "okay";
```
What about these?
```
	marvell,xenon-emmc;
	marvell,xenon-tun-count = <9>;

	pinctrl-names = "default";
	pinctrl-0 = <&mmc_pins>;

	#address-cells = <1>;
	#size-cells = <0>;
	mmccard: mmccard@0 {
		compatible = "mmc-card";
		reg = <0>;
	};
```

dhewg commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dts:
```diff
@@ -0,0 +1,236 @@
+	usb3_phy: usb3-phy {
+		compatible = "usb-nop-xceiv";
+		vcc-supply = <&reg_usb3_vbus>;
+	};
```
I see you removed the `&usb3` node, which makes these 2 blocks unused now. Maybe remove them too and then add everything usb3 related once it's working?
(With the previous `&usb3` node, you mixed another phy setup with the comphy setup from `armada-37xx.dtsi`, which may be related to the problem)

dhewg reviewed (commented):
Some more comments, just by comparing the non-ultra dts to this one.

v-vid commented:
Hi @dhewg, sorry for the late reply. Got caught in something else..

Did some digging and found that the predinstalled Marvell u-boot is the root cause of the issues.
After adding Ultra support in u-boot and ATF packages and flashing 2020.10-rc4 on device, both USB and wifi issues were resolved and everything seems to be working now.
Reference: https://lore.kernel.org/linux-arm-kernel/!&!AAAAAAAAAAAuAAAAAAAAAOBWTR25SONAuESb5loyl%2FsBAMO2jhD3dRHOtM0AqgC7tuYAAAAAAA4AABAAAADpLpjRTgFBSZBSAreF%2FAshAQAAAAA=@stahurabrenner.com/T/

If it's okay to you, I would like to revert the last dts changes as everything works on 4.19 as it is, and by flashing the new u-boot it also works on 5.4.

I'll create a new PR for u-boot and ATF when I clean up the patches.

dhewg commented:
Sure, but I'm not a commiter, so it's not my call anyway :)
But please just post the ultra dts for mainline review, it'll make any further work so much easier for everyone involved!

v-vid commented:
> Sure, but I'm not a commiter, so it's not my call anyway :)
> But please just post the ultra dts for mainline review, it'll make any further work so much easier for everyone involved!

fair enough. Cleaned up dts and sent PR to mainline a couple of minutes ago, also updated the dts in this PR.

hauke commented:
Some backports are merged to master, could you please try if USB and PCIe are now working with kernel 5.4

dhewg commented on target/linux/mvebu/image/cortexa53.mk:
```diff
@@ -30,6 +30,16 @@ define Device/globalscale_espressobin-emmc
+  DEVICE_MODEL := ESPRESSObin
+  DEVICE_VARIANT := Ultra
+  DEVICE_PACKAGES += kmod-i2c-pxa kmod-rtc-pcf8563
+  SOC := armada-3720
```
with #3405 merged you can now add
`BOOT_SCRIPT := espressobin`

v-vid replied:
added in the last change, also updated uboot-envtools as discussed earlier.

v-vid commented:
> Some backports are merged to master, could you please try if USB and PCIe are now working with kernel 5.4

Hey Hauke, the issues on 5.4 were already resolved by updating the u-boot/ATF firmware.

adschm commented:
The last push seems to be an older version?

v-vid commented:
> The last push seems to be an older version?

@adschm sorry, looks like I've mixed the previous commit somehow. I've updated the device tree based on the feedback from the mainline linux PR and fixed all regression from the previous commit. It's ready for merge now.

adschm commented on package/boot/uboot-envtools/files/mvebu:
```diff
@@ -26,6 +26,7 @@ glinet,gl-mv1000)
 	;;
 globalscale,espressobin|\
 globalscale,espressobin-emmc|\
+globalscale,espressobin-ultra|\
```
Technically, you should bump PKG_RELEASE for uboot-envtools when changing the content.

v-vid replied:
Missed that one, should I split the this part and make a separate PR with PKG_RELEASE bump?
From what I can see, that's how it's usually done but would just like to check first.

adschm replied:
Not necessarily, that's essentially matter of taste, different people do it differently. As you prefer ...

adschm commented on target/linux/mvebu/files/arch/arm64/boot/dts/marvell/armada-3720-espressobin-ultra.dts:
```diff
@@ -0,0 +1,241 @@
+		vcc-supply = <&reg_usb3_vbus>;
+	};
+
+	gpio-leds {
```
Just "leds"

v-vid replied:
updated.

adschm commented on target/linux/mvebu/image/cortexa53.mk:
```diff
@@ -32,6 +32,17 @@ define Device/globalscale_espressobin-emmc
 
+define Device/globalscale_espressobin-ultra
+  $(call Device/Default-arm64)
+  DEVICE_VENDOR := Marvell
```
I don't understand why vendor is Marvell for this variable and globalscale for the image, but that is consistently inconsistent here, so probably we shouldn't start a new battle here.

v-vid replied:
Doesn't make much sense to me either to be honest, but I was just following convention.

adschm commented:
Seems good now on a quick look, haven't checked the DTS in detail though (apart from my few nitpick comments).

ynezz commented:
Thanks! Pulled into my staging tree at https://git.openwrt.org/openwrt/staging/ynezz.git
